Xamarin.Forms.Platform.Android/Renderers/EntryRenderer.cs

```diff
@@ -281,45 +294,123 @@ void UpdateReturnType()
+			if (!_cursorPositionChangePending)
+			{
+				var start = cursorPosition;
 
-			var selectionLength = control.SelectionEnd - control.SelectionStart;
-			if (selectionLength != Element.SelectionLength)
-				ElementController?.SetValueFromRenderer(Entry.SelectionLengthProperty, selectionLength);
-		}
+				if (selectionStart != start)
+					SetCursorPositionFromRenderer(selectionStart);
+			}
 
+			if (!_selectionLengthChangePending)
+			{
+				int elementSelectionLength = System.Math.Min(Control.Text.Length - cursorPosition, Element.SelectionLength);
+
+				var controlSelectionLength = Control.SelectionEnd - selectionStart;
+				if (controlSelectionLength != elementSelectionLength)
+					SetSelectionLengthFromRenderer(controlSelectionLength);
+			}
+		}
 
 		void UpdateCursorSelection()
 		{
-			var control = Control;
-			if (control == null || Element == null)
+			if (_nativeSelectionIsUpdating || Control == null || Element == null)
 				return;
 
-			if (Element.IsSet(Entry.CursorPositionProperty) || Element.IsSet(Entry.SelectionLengthProperty))
+			if (Control.RequestFocus())
 			{
-				var start = Element.CursorPosition;
-				var end = System.Math.Min(control.Length(), Element.CursorPosition + Element.SelectionLength);
+				try
+				{
+					int start = GetSelectionStart();
+					int end = GetSelectionEnd(start);
 
-				if (control.SelectionStart != start || control.SelectionEnd != end)
+					Control.SetSelection(start, end);
+				}
+				catch (System.Exception ex)
 				{
-					control.SetSelection(start, end);
-					control.RequestFocus();
+					Internals.Log.Warning("Entry", $"Failed to set Control.Selection from CursorPosition/SelectionLength: {ex}");
 				}
+				finally
+				{
+					_cursorPositionChangePending = _selectionLengthChangePending = false;
+				}
+			}
+		}
+
+		int GetSelectionEnd(int start)
+		{
+			int end = start;
+			int selectionLength = Element.SelectionLength;
+
+			if (Element.IsSet(Entry.SelectionLengthProperty))
+				end = System.Math.Max(start, System.Math.Min(Control.Length(), start + selectionLength));
+
+			int newSelectionLength = System.Math.Max(0, end - start);
+			if (newSelectionLength != selectionLength)
+				SetSelectionLengthFromRenderer(newSelectionLength);
+
+			return end;
+		}
+
+		int GetSelectionStart()
+		{
+			int start = Control.Length();
+			int cursorPosition = Element.CursorPosition;
+
+			if (Element.IsSet(Entry.CursorPositionProperty))
+				start = System.Math.Min(Control.Text.Length, cursorPosition);
+
+			if (start != cursorPosition)
+				SetCursorPositionFromRenderer(start);
+
+			return start;
+		}
+
+		void SetCursorPositionFromRenderer(int start)
+		{
+			try
+			{
+				_nativeSelectionIsUpdating = true;
+				ElementController?.SetValueFromRenderer(Entry.CursorPositionProperty, start);
+			}
+			catch (System.Exception ex)
+			{
+				Internals.Log.Warning("Entry", $"Failed to set CursorPosition from renderer: {ex}");
+			}
+			finally
+			{
+				_nativeSelectionIsUpdating = false;
```

Comment:
In what case is the finally needed here? This is confusing.

Reply:
@PureWeen wanted us to be sure we didn't get stuck in a non-operational state in the event we're unable to get or set the value.